### DISCUSSION TITLE
Remove DictionaryExtensions ?

### DISCUSSION THREAD

**Discussion by drieseng** [quartznet/quartznet#1382] [Ideas] [answered]
We currently have a **DictionaryExtensions** class with a `TryGetAndReturn(...)` convenience method.
I don't think this class should have been public, as I don't think the goal of this project is to deliver extension methods for **IDictionary<TKey, TValue>**.

Apart from that:
* This extension method is -  performance wise - not efficient. Even if you use it on a concrete **Dictionary<TKey, TValue>** instance this will (may?) result in an virtual method call for `IDictionary<TKey, TValue>.TryGetValue(...)`. I doubt the JIT will be able to devirtualize this call since **Dictionary<TKey, TValue>** is not sealed.
* You lose the information that the out value cannot be null when `Dictionary<TKey, TValue>.TryGetValue(...)` returns true.

I propose to at least mark this class internal, and preferrably remove this class and its usage altogether.

@lahma, what do you think?

**Comment by lahma** [comment, marked as answer]
This is a broader issue. There are a lot of classes that should have been internal from the get-go, like ObjectExtensions. I welcome a PR that hides internal-looking classes like these. 

Having virtual members has been adaptation from Java where everything's virtual by default.
